src/compute/src/communication.rs

```diff
@@ -0,0 +1,230 @@
+// the connection must receive it in order to consider the connection to be
+// successfully established.
+const ACK_MAGIC: u64 = 0x68656c6f77726c64;
+
+/// Fork of timely::execute::execute that calls into our fork of create_sockets
+pub fn execute_acking_connections<T, F>(
+    mut config: Config,
+    func: F,
+) -> Result<WorkerGuards<T>, String>
+where
+    T: Send + 'static,
+    F: Fn(&mut Worker<timely::communication::Allocator>) -> T + Send + Sync + 'static,
+{
+    if let CommunicationConfig::Cluster { ref mut log_fn, .. } = config.communication {
+        *log_fn = Box::new(|events_setup| {
+            let mut result = None;
+            if let Ok(addr) = ::std::env::var("TIMELY_COMM_LOG_ADDR") {
+                use timely::dataflow::operators::capture::EventWriterCore;
+                use timely::logging::BatchLogger;
+
+                eprintln!("enabled COMM logging to {}", addr);
+
+                if let Ok(stream) = TcpStream::connect(&addr) {
+                    let writer = EventWriterCore::new(stream);
+                    let mut logger = BatchLogger::new(writer);
+                    result = Some(timely::logging_core::Logger::new(
+                        ::std::time::Instant::now(),
+                        ::std::time::Duration::default(),
+                        events_setup,
+                        move |time, data| logger.publish_batch(time, data),
+                    ));
+                } else {
+                    panic!("Could not connect to communication log address: {:?}", addr);
+                }
+            }
+            result
+        });
+    }
+
+    let (allocators, other) = match config.communication {
+        timely::CommunicationConfig::Cluster {
+            threads,
+            process,
+            addresses,
+            report,
+            log_fn,
+        } => {
+            let sockets_result = create_sockets(addresses, process, report);
+            let result: Result<(Vec<GenericBuilder>, Box<dyn Any + Send>), String> =
+                match sockets_result.and_then(|sockets| {
+                    initialize_networking_from_sockets(sockets, process, threads, log_fn)
+                }) {
+                    Ok((stuff, guard)) => Ok((
+                        stuff.into_iter().map(GenericBuilder::ZeroCopy).collect(),
+                        Box::new(guard),
+                    )),
+                    Err(err) => Err(format!("failed to initialize networking: {}", err)),
+                };
+            result
+        }
+        _ => config.communication.try_build(),
+    }?;
+
+    let worker_config = config.worker;
+    initialize_from(allocators, other, move |allocator| {
+        let mut worker = Worker::new(worker_config.clone(), allocator);
+
+        // If an environment variable is set, use it as the default timely logging.
+        if let Ok(addr) = ::std::env::var("TIMELY_WORKER_LOG_ADDR") {
```

Comment:
Same thing with this conditional branch; I think MZ abhors this, and tbh I hadn't realized it was still in Timely (vs packed into the `WorkerConfig` struct).